feat: adds code blocks

ghgoodreau opened the pull request:
Added support for code blocks in my fork, figured they might be wanted here! Feels like a good improvement, open to suggestions of course.

Here's a quick demo of how it works:

https://user-images.githubusercontent.com/25194960/228704711-12867cec-3b6e-4ba1-932c-37b64419ad9c.mp4

Let me know what you think!

transmissions11 commented:
oo @matthuang will be very happy w/ this, he requested this feature a while ago hehe

fantastic job on the first pass, the needed addition of an edit button is a lil annoying but had to be done ig— dont see a way around it

can we maybe make it just a lil icon that only shows up on hover like so:

<img width="678" alt="Screenshot 2023-03-30 at 12 44 54 AM" src="https://user-images.githubusercontent.com/26209401/228773396-d018c086-fbda-408d-9fd2-49ab4625b278.png">

also i think the copy code button is a little too in the way rn, but not sure the best place to put it, maybe we just ripoff chatgpt here?

<img width="453" alt="Screenshot 2023-03-30 at 1 16 45 AM" src="https://user-images.githubusercontent.com/26209401/228773803-885de53d-1c5f-42e7-8f8b-dc2a87a03f80.png">

ghgoodreau commented:
![Screenshot 2023-03-30 at 11 15 54 AM](https://user-images.githubusercontent.com/25194960/228899891-3735d2bb-cd45-4432-b3ef-0b8b231444ef.png)

Thanks for the suggestions! I'm working on these today. @transmissions11, here's what I'm working with so far for the new copy button - does Paradigm have an opinion on which icon packs to use? I'm currently just keeping the buttons as text for now until I have some direction there.

Stepping away for a few, but when I'm back I'll look into the new edit button and stopping generation.

transmissions11 commented:
>  does Paradigm have an opinion on which icon packs to use?

chakra ui comes bundled w/ some icons and happens to have the same perfect edit button so lets use that if possible! https://chakra-ui.com/docs/components/icon

transmissions11 commented on src/components/utils/TextAndCodeBlock.tsx:
```diff
@@ -0,0 +1,107 @@
+          {code}
+        </SyntaxHighlighter>
+      </Box>
+      <TextAndCodeBlock text={after} />
```
oo interesting to go recursive here, do we know if this has a big impact on performance vs a loop?

ghgoodreau replied:
I'm pretty sure the perf is roughly the same for this use case - there isn't a massive advantage one way or another. Would you prefer I used iteration for maintenance/readability purposes?

transmissions11 replied:
hmmm yea i think so, tho nw if it causes too much trouble :)

transmissions11 replied:
sgtm!

transmissions11 commented on package.json:
```diff
@@ -23,6 +23,8 @@
     "react-beforeunload": "^2.5.3",
     "react-dom": "^18.2.0",
     "react-hotkeys-hook": "^4.3.7",
+    "react-markdown": "^8.0.6",
```
this doesnt seem to be used anywhere atm

ghgoodreau commented:
https://user-images.githubusercontent.com/25194960/228968379-b66ac1ef-399a-41d1-94fa-b92f7fa5cff7.mp4

@transmissions11 addressed your comments and changed the edit button according to your suggestion.

Do we want the edit button on user nodes too? I left it out just because that's probably the most commonly edited area and most people won't be formatting code in their prompts, but maybe it will be good for consistent UX. Quick change if we want to add it.

h0tw4t3r commented on src/components/Prompt.tsx:
```diff
@@ -92,6 +109,13 @@ export function Prompt({
               borderRadius="6px"
               borderLeftWidth={isLast ? "4px" : "0px"}
+              position="relative"
+              onMouseEnter={() =>
+                !(data.fluxNodeType === FluxNodeType.User) && setHoveredNodeId(node.id)
+              }
+              onMouseLeave={() =>
+                !(data.fluxNodeType === FluxNodeType.User) && setHoveredNodeId(null)
+              }
```
Two logical operators instead of one?
```suggestion
              onMouseEnter={() =>
                data.fluxNodeType !== FluxNodeType.User && setHoveredNodeId(node.id)
              }
              onMouseLeave={() =>
                data.fluxNodeType !== FluxNodeType.User && setHoveredNodeId(null)
              }
```

ghgoodreau replied:
Good catch. Fixed this and moved copyToClipboard to a util file

h0tw4t3r commented on src/components/utils/TextAndCodeBlock.tsx:
```diff
@@ -0,0 +1,107 @@
+const CopyCodeButton = ({ code }: { code: string }) => {
+  const [copied, setCopied] = useState(false);
+
+  const copyToClipboard = async () => {
+    try {
+      await navigator.clipboard.writeText(code);
+      setCopied(true);
+    } catch (err) {
+      console.error("Failed to copy to clipboard", err);
+    }
+  };
```
Could be moved to `utils`?

h0tw4t3r reviewed (commented):
Some thoughts.

transmissions11 commented on src/utils/clipboard.ts:
```diff
@@ -0,0 +1,9 @@
+export const copyToClipboard = async (code: string): Promise<boolean> => {
+  try {
+    await navigator.clipboard.writeText(code);
+    return true;
+  } catch (err) {
+    console.error("Failed to copy to clipboard", err);
+    return false;
+  }
+};
```
lets use this here too? https://github.com/transmissions11/flux/blob/1ccbfd3e7cd8d469526dd69754be8059f7e866c9/src/components/App.tsx#L759

ghgoodreau replied:
Moved copyMessagesToClipboard to /utils/clipboard.ts

transmissions11 commented:
> addressed your comments and changed the edit button according to your suggestion.

looks fantastic!

> Do we want the edit button on user nodes too? I left it out just because that's probably the most commonly edited area and most people won't be formatting code in their prompts, but maybe it will be good for consistent UX. Quick change if we want to add it.

yeah i think so now, sets us up nicely for https://github.com/transmissions11/flux/issues/24

transmissions11 commented on src/components/Prompt.tsx:
```diff
@@ -118,31 +142,57 @@ export function Prompt({
                 </Center>
               ) : (
                 <>
+                  <Button
+                    display={hoveredNodeId === node.id ? "block" : "none"}
+                    onClick={() => setIsEditing(!isEditing)}
+                    position="absolute"
+                    top={1}
+                    right={1}
+                    zIndex={10}
+                    variant="outline"
+                    border="0px"
+                    p="1"
+                  >
+                    {(isEditing && hoveredNodeId === node.id) ? <CheckIcon boxSize={4} /> : <EditIcon boxSize={4} />}
+                  </Button>
```
<img width="542" alt="Screenshot 2023-03-30 at 3 52 52 PM" src="https://user-images.githubusercontent.com/26209401/228982258-bbd81f5c-1b5c-4dfb-99e2-d32a254d04aa.png">

hmmmm even tho this is position=absolute it seems to be on the same plane as the textbox?

transmissions11 replied:
also lets get rid of the grey background on hover

ghgoodreau replied:
> hmmmm even tho this is position=absolute it seems to be on the same plane as the textbox?

Not 100% sure what you mean, but I added a margin to the textbox so the edit button doesn't overlap with any underlying text. If we are ok with it overlapping I can remove the margin

transmissions11 replied:
ah ok yeah hmmm

like ideally we could just add a margin at the lil corner it occupies and not the whole right side but seems difficult

maybe lets just shrink that margin a bit? feels a little excessive atm

transmissions11 replied:
e.g. lets get rid of this red space

![Uploading image.png…]()

transmissions11 replied:
e.g. lets get rid of this red space

![image](https://user-images.githubusercontent.com/26209401/228985960-e481f1f5-7514-4c0e-b479-832371cc921c.png)

ghgoodreau replied:
Done!

transmissions11 commented on src/components/Prompt.tsx:
```diff
@@ -55,6 +55,10 @@ export function Prompt({
 
   const textOffsetRef = useRef<number>(-1);
 
+  useEffect(() => {
+    setIsEditing(promptNodeType === FluxNodeType.User);
+  }, [promptNode.id]);
+
```
can combine this with the other useeffect that depends on promptNode.id (getting nitty at this point so im happy to handle this part)

ghgoodreau replied:
Went ahead and handled this since I was already in the file

transmissions11 commented:
<img width="680" alt="Screenshot 2023-03-30 at 7 55 51 PM" src="https://user-images.githubusercontent.com/26209401/229011359-47dd4cec-a33b-4df4-824c-53785c947fab.png">

seeing this for long lines, looks like we need a `break-words` or something somewhere?

ghgoodreau commented:
> @transmissions11 seeing this for long lines, looks like we need a `break-words` or something somewhere?

added break word to the parent column seems to have fixed this, think this was introduced sometime when fixing comments

h0tw4t3r commented on src/components/Prompt.tsx:
```diff
@@ -57,16 +69,29 @@ export function Prompt({
-    promptBox?.focus();
+    // TODO: Really wish we didn't have to do this...
+    // is there an optimization we can make somewhere?
+    // Need a small timeout to ensure the text box is rendered.
+    setTimeout(() => {
+      const promptBox = window.document.getElementById(
+        "promptBox"
+      ) as HTMLTextAreaElement | null;
 
-    // Default to moving to the start of the text.
-    textOffsetRef.current = -1;
+      // Focus the text box and move the cursor to chosen offset (defaults to end).
+      promptBox?.setSelectionRange(textOffsetRef.current, textOffsetRef.current);
+      promptBox?.focus();
+
+      // Default to moving to the start of the text.
+      textOffsetRef.current = -1;
+    }, 50);
```
I guess that you need to compose `TextareaAutosize` in another component, let's say `PromptBox`, and use `useEffect` inside it, so you ensure the effect triggers specifically on this text area mount.

It's unfortunate that `react-textarea-autosize` library doesn't handle refs...

transmissions11 replied:
ah i figured out the issue, its not that there's any lag in rendering its cuz isEditing is set in the useEffect is what actually renders the textbox... fixing

h0tw4t3r commented on src/components/utils/TextAndCodeBlock.tsx:
```diff
@@ -0,0 +1,110 @@
+};
+
+export const TextAndCodeBlock = memo(({ text }: { text: string }) => {
+  const match = getNextCodeBlockMatch(text);
```
Can this operation be hard to perform? Since we could have a large prompt, (32k tokens with gpt4 right?), and this being matched on every render, we could memoize it.

transmissions11 replied:
the whole component is memoized — so another memo wouldnt do much i believe

h0tw4t3r reviewed (commented):
Just a few comments!

h0tw4t3r commented:
@transmissions11 small off-topic, what does `nit` correspond to in conventional commits? I see it for the first time...

transmissions11 commented on src/components/utils/TextAndCodeBlock.tsx:
```diff
@@ -0,0 +1,111 @@
+    const match = CODE_BLOCK_DETECT_REGEX.exec(remainingText);
+
+    if (!match) {
+      elements.push(<Text key={elements.length}>{remainingText}</Text>);
+      break;
+    }
+
+    const before = remainingText.substring(0, match.index);
+    const language =
+      CODE_BLOCK_LANGUAGE_DETECT_REGEX.exec(match[1])?.[0]?.substring(3) || "plaintext";
+    const code = match[2].trim();
+    const after = remainingText.substring(match.index + match[0].length);
+
+    if (before.length > 0)
+      elements.push(
+        <Text key={elements.length} mb={4}>
+          {before}
+        </Text>
+      );
+
+    elements.push(
+      <Box key={elements.length} borderRadius="4px" overflow="hidden">
+        <TitleBar language={language} code={code} />
+        <SyntaxHighlighter
+          language={language}
+          wrapLongLines
+          style={coy}
+          codeTagProps={{ style: { wordBreak: "break-word" } }}
+          customStyle={{ padding: "10px", margin: "0px", borderRadius: "0 0 4px 4px" }}
+        >
+          {code}
+        </SyntaxHighlighter>
+      </Box>
+    );
+
+    remainingText = after;
+  }
+
+  return <Box width="100%">{elements}</Box>;
+});
```
this being recursive was starting to eat me up inside hehe— the thought of the extra function calls was keeping me up at night

@ghgoodreau does this still look ok? i used gpt-4 to rewrite it iteratively lolol

ghgoodreau replied:
This looks good to me! I'm not around to test it atm but if the functionality is the same I'd say it's ready to merge.

And no worries @transmissions11 Appreciate the thoughtful reviews. I'm here to improve and build

transmissions11 reviewed (approved):
approved given @ghgoodreau approves of the iterative solution

transmissions11 commented:
LFG tysm again @ghgoodreau and everyone else involved
